File: backend/app/models/prediction.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime

class LocationInput(BaseModel):
    latitude: float
    longitude: float
    location_name: str
    radius_km: int = 100

class EngineResult(BaseModel):
    engine_type: str
    location: LocationInput
    predictions: List[Dict[str, Any]]
    summary: Dict[str, Any]
    processing_time: float
    timestamp: datetime

class CombinedPrediction(BaseModel):
    location: LocationInput
    brett_earth_result: Optional[EngineResult] = None
    brett_space_result: Optional[EngineResult] = None
    combined_summary: Dict[str, Any]
    timestamp: datetime

class CymaticData(BaseModel):
    location: LocationInput
    wave_field: List[List[List[float]]]
    frequency_data: Dict[str, Any]
    visualization_metadata: Dict[str, Any]
    timestamp: datetime

class DataSourceStatus(BaseModel):
    source_name: str
    status: str
    last_update: Optional[str] = None
    error_message: Optional[str] = None
    reliability_percent: float

class SystemStatus(BaseModel):
    operational: bool
    data_sources: List[DataSourceStatus]
    last_refresh: Optional[datetime] = None
    next_refresh: Optional[datetime] = None
    error_count: int = 0

class PredictionDay(BaseModel):
    day: int
    date: str
    magnitude_prediction: float
    probability_percent: float
    risk_level: str
    confidence: float

class PredictionSummary(BaseModel):
    location: LocationInput
    engine_type: str
    max_magnitude: float
    highest_risk_day: int
    average_probability: float
    total_predictions: int
    risk_distribution: Dict[str, int]

File: frontend/src/types/prediction.ts
export interface Prediction {
  day: number;
  date: string;
  magnitude_prediction: number;
  magnitude_estimate: number;
  probability_percent: number;
  risk_level: string;
  confidence: number;
  confidence_level: string;
}

export interface LocationInput {
  latitude: number;
  longitude: number;
  location_name: string;
  radius_km: number;
}

export interface EngineResult {
  engine_type: string;
  location: LocationInput;
  predictions: Prediction[];
  summary: {
    max_magnitude: number;
    highest_risk_day: number;
    average_probability: number;
    total_predictions: number;
    risk_distribution: {
      [key: string]: number;
    };
  };
  processing_time: number;
  timestamp: string;
}

export interface CombinedPrediction {
  location: LocationInput;
  brett_earth_result?: EngineResult;
  brett_space_result?: EngineResult;
  combined_summary: {
    consensus_max_magnitude: number;
    consensus_highest_risk_day: number;
    combined_confidence: number;
    engine_agreement_percent: number;
  };
  timestamp: string;
}
